library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,152 @@
+have changed, there is potentially a very high cost to this option, which should
+be weighed carefully. Case 1 (neither renaming the Maven ID nor the
+Java package) doesn't have this code updating cost, and the fewer the breakages,
+the easier the transition is between major versions. Basically, the cost of
+diamond dependency conflicts has to be weighed against the cost of updating
+import statements everywhere the library is used.
+
+Let's take examples from two extremes.
+
+1. A library with 10,000 classes, and 1 function is deprecated between major
+   version 1 to major version 2, and it is used in one place in a large
+   dependency tree.
+
+In this case, moving 10,000 classes to a new package in a large dependency tree
+would be a very expensive endeavor. In contrast, updating the one place where
+the old function is used to use the new function instead would be considerably
+less work and could be rolled out much more quickly. In this scenario, it is
+clearly superior to keep the same Maven ID and Java package.
+
+3. A library with 1,000 classes, and a large refactoring touching 750 of the
+   classes is done between major version 1 and major version 2.
+
+In this case, changing consuming code would be a large undertaking, and it's not
+certain that all consuming code would feel it's worth it to migrate to the new
+major version. If the library author decided to opt for keeping the same Maven
+ID and same Java package, the ecosystem would need to bifurcate in
+order to handle code paths requiring one versus the other major version. Either
+the ecosystem would generally opt to retain the old major version, or there
+would be an extended period of difficult diamond dependency conflicts before
+everyone had transitioned. In this scenario, it is clearly superior to rename
+both the Maven ID and Java package.
+
+Given these considerations, the rule to follow can be summed up as:
+
+*When making breaking changes between major versions, prefer to rename both the
+Maven ID and the Java package. However, if the ratio of surface
+breakage to library size is sufficiently low and the number of usage points of
+breaking code is sufficiently low, keeping the same Maven ID and Java
+package can be preferred instead. Under no circumstance should only one of the
+two be renamed without the other being renamed also.*
+
+Example renames in open source 
+[Square has established it as a policy](http://jakewharton.com/java-interoperability-policy-for-major-version-updates/)
+for its Java libraries (examples include OkHttp and Retrofit). Numerous open
+source packages have performed such a rename between large major version
+updates:
+- OkHttp (com.squareup.okhttp -> com.squareup.okhttp3)
+- Apache Commons Lang (org.apache.commons.lang -> org.apache.commons.lang3)
+- RxJava (rx (version 1.x) -> io.reactivex (version 2.x))
+- JDOM (org.jdom -> org.jdom2)
+- JUnit (junit.framework (versions 1.x-3.x) -> org.junit (version 4))
```

Comment:
This may be wrong. JUnit 4 is case 1. They did not change the package name and did not change the artifact and group IDs. They *added* a new org.junit package but kept the old junit.framework package too.

JUnit 5 is more complicated, and arguably something we don't consider here. They've now broken the product into several jars and multiple artifacts with new IDs and packages. However they do have a junit-vintage artifact that includes backwards compatibility.

Reply:
For JUnit 4, if the "new" JUnit provided was under a new package name (`org.junit`), I would argue that they did a rename of the package. The fact that the old `junit.framework` was also included seems like a distribution strategy to me (it's at the packaging level, not the code level). Thus, I think case 1 is ruled out for JUnit. 

I think the text of case 3 includes the JUnit scenario (at least I tried to make it do so), but I think now that it's potentially confusing. I think I'll create a case 5 for it (rename Java package, bundle old & new in the same maven ID) so that it's clearer.